codec-http/src/main/java/io/netty/handler/codec/http/router/OrderlessRouter.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.router;
+
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Router that doesn't contain information about HTTP request methods and route
+ * matching orders.
+ */
+class OrderlessRouter<T> {
+    // A path can only point to one target
+    private final Map<Path, T> routes = new HashMap<Path, T>();
+
+    // Reverse index to create reverse routes fast (a target can have multiple paths)
+    private final Map<T, Set<Path>> reverseRoutes = new HashMap<T, Set<Path>>();
+
+    //--------------------------------------------------------------------------
+
+    public Map<Path, T> routes() {
+        return routes;
+    }
+
+    /**
+     * This method does nothing if the path has already been added.
+     * A path can only point to one target.
+     */
+    public OrderlessRouter<T> addRoute(String path, T target) {
+        Path p = new Path(path);
+        if (routes.containsKey(path)) {
+            return this;
+        }
+
+        routes.put(p, target);
+        addReverseRoute(target, p);
+        return this;
+    }
+
+    private void addReverseRoute(T target, Path path) {
+        Set<Path> paths = reverseRoutes.get(target);
+        if (paths == null) {
+            paths = new HashSet<Path>();
+            paths.add(path);
+            reverseRoutes.put(target, paths);
+        } else {
+            paths.add(path);
+        }
+    }
+
+    //--------------------------------------------------------------------------
+
+    /** Removes the route specified by the path. */
+    public void removePath(String path) {
+        Path p      = new Path(path);
+        T    target = routes.remove(p);
```

Comment:
It's column aligned.

![column-aligned](https://cloud.githubusercontent.com/assets/10095/8274162/cb0d2ee4-18ce-11e5-875d-525101caacfb.png)